File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        boolean direction = true;
        Queue<TreeNode> que = new ArrayDeque<>();
        que.add(root);

        while (!que.isEmpty()) {
            int len = que.size();
            List<Integer> temp = new ArrayList<>();

            for (int i = 0; i < len; i++) {
                TreeNode node = que.poll();
                temp.add(node.val);

                if (node.left != null) {
                    que.add(node.left);
                }
                if (node.right != null) {
                    que.add(node.right);
                }
            }

            if (direction) {
                result.add(temp);
                direction = false;
            } else {
                Collections.reverse(temp);
                result.add(temp);
                direction = true;
            }
        }

        return result;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
from collections import deque


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = list()
        if root is None: return result

        direction = True
        que = deque([root])

        while que:
            length = len(que)
            temp = list()

            for i in range(length):
                node = que.popleft()
                temp.append(node.val)

                if node.left: que.append(node.left)
                if node.right: que.append(node.right)

            if direction:
                result.append(temp)
                direction = False
            else:
                result.append(temp[::-1])
                direction = True

        return result